### PR TITLE
Native application feature

### PR DESCRIPTION
**Pull request by sihil**
The native application feature provides an easy way to download, install and start a package that was generated by the sbt-native-packager tool.

Instead of the large amounts of cfn-init metadata and the many lines of user data, a typical application can now have a single line to do the right thing using convention. For example:

```
/opt/feature/native-packager/install.sh -b composer-dist -s
```

Here we are telling it to download from the `composer-dist` bucket and automatically start the application. In order to get to that point the script will:
- Retrieve the tags of the instance
- Create a user (using the value of the Stack tag as the username)
- Download the package (using the bucket and value of the Stack, Stage and App tags to determine the location - assumes that it is a `.tar.gz`)
- Expand the package into the user home directory
- Create an upstart script using a sane template
- Start the upstart service

It's not perfect, but I think will save quite a lot of time in a process that is quite error prone. Am interested in feedback from others in the team.

### CODE REVIEW

**Comment by sihil** [comment]
Also, fun fact: did you know bash had associative arrays?

**Comment by sihil** [comment]
Any thoughts @lindseydew; @theefer; @kenoir; @cb372 ?

**Comment by theefer** [comment]
This looks like it'd remove some boilerplate from [our CloudFormation scripts](https://github.com/guardian/grid-infra/blob/master/platform/cloud-formation/media-service.json#L802) and [our individual projects](https://github.com/guardian/media-service/blob/master/media-api/conf/media-api.conf) so that's a good move imho. I like things that work out of the box and devs don't have to worry about.

Note that the Grid currently uses the mega-JAR approach so we'd need to migrate to the more recent native-package/ZIP setup, but that's something we've wanted to do anyway.

**Comment by theefer** [comment]
Grid change now tracked at https://github.com/guardian/media-service/issues/749.

File: packer/resources/features/native-packager/install.sh [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+STAGE=$(sub "tag.Stage")
+APP=$(sub "tag.App")
+
+if [ -z "${USER}"]; then
+  USER=${STACK}
+fi
+HOME_DIR="/home/${USER}"
+
+# Make user
+if [ "${USER}" != "${DEFAULT_USER}" ]; then
```
Comment (cb372) [inline packer/resources/features/native-packager/install.sh]: Where is this `DEFAULT_USER` defined?

**Comment by cb372** [comment]
This looks really good. I think we can use it for some of our CAPI services.

Going off-topic slightly, I'm a little bit worried that we're reaching the level of complexity where bash scripts become more trouble than they're worth. I'd be interested in using something a little more feature-rich and composable for provisioning, rather than re-inventing Puppet in bash.

 e.g.
- You had to roll your own templating system, whereas Chef/Puppet/Ansible provide this for free
- One of your scripts includes "create a user if it doesn't already exist" logic, which is likely to be copy-pasted into a number of different features (e.g. I want it for the Logstash feature I'm writing). It would be nice if this was defined as a recipe step that could be referenced rather than copy-pasted.
- Once the number of features increases and dependencies inevitably develop between them, managing these dependencies will become quite tricky without a provisioning tool to take care of it for us

Personally I'm a fan of Ansible for provisioning, although it doesn't integrate very nicely with Packer.

File: packer/resources/features/native-packager/install.sh (hunk 1 above)
Comment (sihil) [inline packer/resources/features/native-packager/install.sh]: It's not. I ripped it out and didn't tidy up. Lucky that USER is never blank :)

**Comment by sihil** [comment]
I totally agree. I'm a big fan of puppet - especially as they are porting it to C++, which should make performance tonnes better soon.

Unfortunately there has been a backlash against puppet as being overly complex and incomprehensible and so people don't use it and I'm fearful of people looking at this repo, seeing puppet code and immediately dismissing it. This comes out of puppet = websys = out of our control and understanding. As a result I'm deliberately sticking with bash so it doesn't gain this association. For now.

If it were to be puppet we would get the templating and user creation stuff for free - and if each feature was modelled as a module, we'd be able to easily declare dependencies and potentially write a wrapper that dealt with the pre-cache stuff we've been discussing.

**Comment by theefer** [comment]
Though I sadly count myself in the group of people scarred and scared by puppet, I did notice that user creation didn't really belong here and would have been best done as an independent step to depend on. So I agree with what you both said and that bash will not scale forever.

**Comment by kenoir** [comment]
@sihil nice. If it's any use `cfn-init` uses [pystache](https://github.com/defunkt/pystache) to do variable substitution.

**Comment by sihil** [comment]
That's interesting @kenoir. I'd happily farm out templates to something like pystache. My version is a bit shonky (e.g. if there is a variable to substitute but no supplied value then there is currently no error).

**Comment by mchv** [comment]
This is interesting!

I have some doubt about the generic `upstart.conf` file, as for instance this is not safe for elasticsearch (see https://github.com/guardian/content-api/commit/0eb3252d4c34e656536ebf07b10f1669c4451a38)

**Comment by sihil** [comment]
@mchv That's fair enough, but I think it can be good enough for 90% of use cases. 

To clarify though, this is a generic upstart script specifically for applications that have been bundled by the sbt-native-packager. I'm not suggesting that it suitable for anything outside of that scope.

We could add a simple way to override the script if required.

**Pull request merged**